File: CRM-Server/app/schemas/lead_ai.py
"""
AI 解析线索信息 Schema

用于 AI 智能创建线索功能的请求和响应结构
"""
from pydantic import BaseModel, Field
from typing import Optional


class LeadAIParseRequest(BaseModel):
    """AI 解析线索信息请求"""
    content: str = Field(..., min_length=1, description="用户输入的自然语言描述")


class LeadAIParsedInfo(BaseModel):
    """AI 解析出的线索信息"""
    lead_name: Optional[str] = Field(None, description="线索名称")
    source: Optional[str] = Field(None, description="线索来源")
    city: Optional[str] = Field(None, description="所在城市")
    company_scale: Optional[str] = Field(None, description="公司规模")
    contact_name: Optional[str] = Field(None, description="联系人姓名")
    contact_phone: Optional[str] = Field(None, description="联系电话")

    # 缺失字段提示
    missing_fields: list[str] = Field(default_factory=list, description="缺失的必填字段")


class LeadAIFollowUpInfo(BaseModel):
    """AI 解析出的跟进记录信息"""
    content: Optional[str] = Field(None, description="跟进内容（除下一步计划外的其他信息）")
    method: Optional[str] = Field(None, description="跟进方式（电话/微信/拜访/邮件）")
    next_action: Optional[str] = Field(None, description="下一步动作/计划")
    next_follow_time: Optional[str] = Field(None, description="下次跟进时间（相对时间表达，如'下周三'、'三天后'）")


class LeadAIParseResponse(BaseModel):
    """AI 解析线索信息响应"""
    lead_info: LeadAIParsedInfo = Field(..., description="解析出的线索信息")
    follow_up_info: Optional[LeadAIFollowUpInfo] = Field(None, description="解析出的跟进记录信息")
    thinking_process: Optional[str] = Field(None, description="AI 思考过程")


class LeadAICreateRequest(BaseModel):
    """AI 创建线索请求（用户确认后提交）"""
    lead_name: str = Field(..., description="线索名称")
    source: str = Field(..., description="线索来源")
    city: str = Field(..., description="所在城市")
    contact_name: str = Field(..., description="联系人姓名")
    contact_phone: str = Field(..., description="联系电话")
    company_scale: Optional[str] = Field(None, description="公司规模")
    # 跟进记录相关
    follow_up_content: Optional[str] = Field(None, description="跟进内容")
    next_action: Optional[str] = Field(None, description="下一步动作")
    next_follow_time: Optional[str] = Field(None, description="下次跟进时间")

File: CRM-Client/src/api/leadAI.ts
/**
 * AI 线索解析 API
 */
import request from '@/utils/request'

export interface LeadAIParseRequest {
  content: string
}

export interface LeadAIParsedInfo {
  lead_name: string | null
  source: string | null
  city: string | null
  company_scale: string | null
  contact_name: string | null
  contact_phone: string | null
  missing_fields: string[]
}

export interface LeadAIFollowUpInfo {
  content: string | null
  next_action: string | null
  next_follow_time: string | null
}

export interface LeadAIParseSSEEvent {
  event: 'status' | 'content' | 'parsed' | 'error'
  message?: string
  content?: string
  lead_info?: LeadAIParsedInfo
  follow_up_info?: LeadAIFollowUpInfo
  thinking_process?: string
}

export interface LeadAICreateRequest {
  lead_name: string
  source: string
  city: string
  contact_name: string
  contact_phone: string
  company_scale?: string
  follow_up_content?: string
  next_action?: string
  next_follow_time?: string
}

export interface LeadAIResult {
  id: number
  lead_name: string
  contact_name: string
  contact_phone: string
  source: string
  city: string
  company_scale?: string
  status: number
  created_time: string
}

export const leadAiApi = {
  /**
   * AI 解析线索信息（SSE 流式响应）
   * @param data 解析请求数据
   * @param onEvent SSE 事件回调
   * @param token JWT token
   */
  parseSSE: async (
    data: LeadAIParseRequest,
    onEvent: (event: LeadAIParseSSEEvent) => void,
    token: string
  ): Promise<void> => {
    const url = '/api/v1/leads/ai/parse'

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify(data)
    })

    if (!response.ok) {
      throw new Error(`HTTP error: ${response.status}`)
    }

    const reader = response.body?.getReader()
    if (!reader) {
      throw new Error('No response body')
    }

    const decoder = new TextDecoder()
    let buffer = ''

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })

      // 解析 SSE 数据
      const lines = buffer.split('\n\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          try {
            const eventData = JSON.parse(line.slice(6)) as LeadAIParseSSEEvent
            onEvent(eventData)

            // 收到 parsed 或 error 事件后结束
            if (eventData.event === 'parsed' || eventData.event === 'error') {
              return
            }
          } catch {
            // 忽略解析错误
          }
        }
      }
    }
  },

  /**
   * 从 AI 解析结果创建线索
   * @param data 创建请求数据
   * @returns 创建成功的线索数据
   */
  createFromAI: (data: LeadAICreateRequest) => {
    return request.post<LeadAIResult>(
      '/v1/leads/ai/create',
      data
    )
  }
}